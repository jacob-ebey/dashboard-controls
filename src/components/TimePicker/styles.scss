@import '../colors.scss';

.highlightBoarder {
  box-shadow: 0 0 3px $color-primary;
  border: 1px solid #0078d7;
}

.errorHighlight {
  box-shadow: 0 0 3px #e81123;
  border: 1px solid #e81123;
  &:focus {
    @extend .errorHighlight;
  }
}

.main {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.error {
  color: #e81123;
  position: absolute;
  display: flex;
  text-align: start;
}

.input {
  box-sizing: border-box;
  border: 1px solid #a6a6a6;
  padding: 5px;
  margin: 0;
  transition: all 0.30s ease-in-out;
  outline: none;
  &:focus {
    @extend .highlightBoarder;
  }
}

.activeItem {
  background-color: #eaeaea;
}

.popup {
  position: absolute;
  display: flex;
  max-height: 200px;
  overflow-y: hidden;
  text-align: start;
  background-color: #ffffff;
  @extend .highlightBoarder;
}

.popupList {
  overflow-y: scroll;
  width: 33.333%;
}

.popupItem {
  display: block;
  padding: 2px;
  cursor: pointer;
  user-select: none;

  &:hover {
    background-color: #eaeaea;
  }
  &:active {
    background-color: #c8c8c8;
  }
}