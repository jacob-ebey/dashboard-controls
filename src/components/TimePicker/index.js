import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import styles from './styles.scss';

function pad(d) {
  return (d < 10) ? '0' + d.toString() : d.toString();
}

const hours = Array(12)
  .fill()
  .map((_, index) => index);
hours[0] = 12;

const minutes = Array(60)
  .fill()
  .map((_, index) => pad(index));

const amPm = ['AM', 'PM'];

export default class TimePicker extends React.PureComponent {
  static propTypes = {
    value: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
      PropTypes.instanceOf(Date),
    ]),
    onChange: PropTypes.func,
  };

  state = {
    popupWidth: undefined,
    isOpen: false,
    internalValue: new Date(),
    internalError: undefined,
  };

  componentDidMount() {
    const { value, onChange } = this.props;

    document.addEventListener('mousedown', this._handleClickOutside);

    const mainNode = ReactDOM.findDOMNode(this.mainRef);
    if (mainNode) {
      const popupWidth = mainNode.getBoundingClientRect().width;
      this.setState({ popupWidth });
    }

    if (!value && onChange) {
      onChange(new Date());
    }

    this.setState({ internalValue: value });
  }

  componentWillReceiveProps(nextProps) {
    const { value } = this.props;

    if (nextProps.value !== value) {
      this.setState({ internalValue: nextProps.value });
    }
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this._handleClickOutside);
  }

  render() {
    const { popupWidth, isOpen, internalError } = this.state;
    const tempDate = this._value;
    const date = typeof tempDate === 'object' ? tempDate : {};
    const formattedDate = typeof tempDate === 'object' ? `${tempDate.hour}:${tempDate.minute} ${tempDate.amPm}` : tempDate;

    return (
      <div
        className={styles.main}
        ref={this._storeMainRef}
      >
        <input
          className={`${styles.input} ${isOpen ? styles.highlightBoarder : ''} ${internalError ? styles.errorHighlight : ''}`}
          onChange={this._handleOnInputChange}
          onClick={this._toggleOpen}
          value={formattedDate}
        />
        {
          !isOpen && internalError && (
            <span
              className={styles.error}
              style={{
                width: popupWidth,
              }}
            >
              {internalError}
            </span>
          )
        }
        {
          isOpen && (
            <div
              className={`${styles.popup} ${internalError ? styles.errorHighlight : ''}`}
              style={{
                width: popupWidth,
              }}
            >
              <div className={styles.popupList}>
                {
                  hours.map(hour => (
                    <span
                      key={hour}
                      className={`${styles.popupItem} ${hour === date.hour ? styles.activeItem : ''}`}
                      onClick={this._onHourClicked(hour)}
                    >
                      {hour}
                    </span>
                  ))
                }
              </div>
              <div className={styles.popupList}>
                {
                  minutes.map(minute => (
                    <span
                      key={minute}
                      className={`${styles.popupItem} ${minute === date.minute ? styles.activeItem : ''}`}
                      onClick={this._onMinuteClicked(minute)}
                    >
                      {minute}
                    </span>
                  ))
                }
              </div>
              <div className={styles.popupList}>
                {
                  amPm.map(option => (
                    <span
                      key={option}
                      className={`${styles.popupItem} ${option === date.amPm ? styles.activeItem : ''}`}
                      onClick={this._onAmPmClicked(option)}
                    >
                      {option}
                    </span>
                  ))
                }
              </div>
            </div>
          )
        }
      </div>
    );
  }

  get _value() {
    const { internalValue } = this.state;

    const date = new Date(internalValue);
    if (!isNaN(date.getHours()) && !isNaN(date.getMinutes())) {
      const pureHours = date.getHours();

      let hours = pureHours;
      if (hours === 0) {
        hours = 12;
      } if (hours > 12) {
        hours = hours - 12;
      }
      const minute = pad(date.getMinutes());

      const tempDate = {
        minute,
        hour: hours,
        amPm: pureHours >= 12 ? 'PM' : 'AM',
      };
      return tempDate;
    }

    return internalValue;
  }

  _handleOnInputChange = event => {
    const { onChange } = this.props;

    const split = event.target.value.split(':');

    if (split.length === 2) {
      const split2 = split[1].split(' ');

      if (split2.length === 2) {
        let hour = parseInt(split[0].trim(), 10);
        const minute = parseInt(split2[0].trim(), 10);
        const amPm = split2[1].trim().toLowerCase();

        if (!isNaN(hour) && hour <= 12 && !isNaN(minute) && minute <= 60 && (amPm === 'am' || amPm === 'pm')) {
          if (amPm === 'am') {
            if (hour === 12) {
              hour = 0;
            }
          } else if (hour !== 12) {
            hour = hour + 12;
          }

          const date = new Date();
          date.setHours(hour, minute);

          if (onChange) {
            onChange(date);
          }

          this.setState({
            internalValue: date,
            internalError: undefined,
          });
          return;
        }
      }
    }

    this.setState({
      internalValue: event.target.value,
      internalError: 'Must be in format 12:00 pm'
    });
  }

  _onHourClicked = hour => () => {
    const { onChange } = this.props;
    const { internalValue } = this.state;

    let date = new Date(internalValue);
    if (!isNaN(date.getTime())) {
      let hours = date.getHours();
      if (!isNaN(hours) && hours >= 12) {
        hours = hour === 12 ? 12 : hour + 12;
      } else {
        hours = hour === 12 ? 0 : hour;
      }

      const minutes = date.getMinutes();

      date.setHours(hours, !isNaN(minutes) ? minutes : 0);
    } else {
      date = new Date();
      date.setHours(hour, 0);
    }

    if (onChange) {
      onChange(date);
    }

    this.setState({
      internalValue: date,
      internalError: undefined,
    });
  }

  _onMinuteClicked = minute => () => {
    const { onChange } = this.props;
    const { internalValue } = this.state;

    let date = new Date(internalValue);
    if (!isNaN(date.getTime())) {
      const hours = date.getHours();
      date.setHours(!isNaN(hours) ? hours : 12, minute);
    } else {
      date = new Date();
      date.setHours(12, minute);
    }

    if (onChange) {
      onChange(date);
    }

    this.setState({
      internalValue: date,
      internalError: undefined,
    });
  }

  _onAmPmClicked = amPm => () => {
    const { onChange } = this.props;
    const { internalValue } = this.state;

    let date = new Date(internalValue);

    if (!isNaN(date.getTime())) {
      let hours = date.getHours();
      const minutes = date.getMinutes();

      if (!isNaN(hours) && !isNaN(minutes)) {
        hours = hours > 12 ? hours - 12 : hours;
        if (amPm === 'AM') {
          date.setHours(hours === 12 ? 0 : hours, minutes);
        } else {
          date.setHours(hours === 12 ? 12 : hours + 12, minutes);
        }
      }
    } else {
      if (amPm === 'AM') {
        date = new Date();
        date.setHours(0, 0);
      } else {
        date = new Date();
        date.setHours(12, 0);
      }
    }

    if (onChange) {
      onChange(date);
    }

    this.setState({
      internalValue: date,
      internalError: undefined,
    });
  }

  _handleClickOutside = event => {
    if (this.mainRef && !this.mainRef.contains(event.target)) {
      this._dismiss();
    }
  }

  _dismiss = event => {
    const { onBlur } = this.props;

    if (onBlur) {
      onBlur(new Event().initEvent('blur', false, false));
    }

    this.setState({ isOpen: false });
  }

  _toggleOpen = () => {
    const { isOpen, onBlur, onFocus } = this.state;

    if (isOpen && onBlur) {
      onBlur(new Event().initEvent('blur', false, false));
    } else if (onFocus) {
      onFocus(new Event().initEvent('focus', false, false));
    }

    this.setState({ isOpen: !isOpen });
  }

  _storeMainRef = mainRef => {
    this.mainRef = mainRef;
  }
}
