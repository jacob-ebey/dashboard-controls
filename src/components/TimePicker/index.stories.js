import React from 'react';
import { storiesOf } from '@storybook/react';

import TimePicker from './index';

class TimePickerWrapper extends React.Component {
  state = {
    value: undefined,
  };

  render() {
    const { value } = this.state;

    return <TimePicker {...this.props} value={value} onChange={this._handleOnChange} />;
  }

  _handleOnChange = value => {
    console.log(value);

    this.setState({ value });
  }
}

storiesOf('TimePicker', module)
  .add('default', () => (
    <TimePickerWrapper />
  ));
