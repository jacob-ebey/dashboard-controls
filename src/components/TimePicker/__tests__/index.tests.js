import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import timemachine from 'timemachine';

import TimePicker from '../index';

Enzyme.configure({ adapter: new Adapter() });

global.window = global;
window.addEventListener = () => { };
window.requestAnimationFrame = () => {
  throw new Error('requestAnimationFrame is not supported in Node');
};

timemachine.config({
  dateString: 'November 30, 1993 1:30:00'
});

describe('components/TimePicker - index.js', () => {
  const createProps = () => ({
    value: new Date(),
    onChange: jest.fn(),
  });

  test('Renders without error', () => {
    const props = createProps();

    const wrapper = shallow(<TimePicker {...props} />);

    expect(props.onChange).toHaveBeenCalledTimes(0);

    expect(wrapper).toMatchSnapshot();
  });

  test('Raises onChange if no value', () => {
    const props = {
      ...createProps(),
      value: '',
    };

    const wrapper = shallow(<TimePicker {...props} />);

    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toHaveBeenCalledWith(new Date());

    expect(wrapper).toMatchSnapshot();
  });

  test('Raises onChange for valid input', () => {
    const props = createProps();

    const date = new Date();
    date.setHours(13, 45);

    const wrapper = shallow(<TimePicker {...props} />);

    const inputOnChange = wrapper.find('input').prop('onChange');
    inputOnChange({ target: { value: '1:45 pm' } });

    wrapper.update();

    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toHaveBeenCalledWith(date);

    expect(wrapper).toMatchSnapshot();
  });

  test('Does not raise onChange for invalid input', () => {
    const props = createProps();

    const wrapper = shallow(<TimePicker {...props} />);

    const inputOnChange = wrapper.find('input').prop('onChange');
    inputOnChange({ target: { value: 'ROFL' } });

    expect(props.onChange).toHaveBeenCalledTimes(0);
  });

  test('Can open dialog', () => {
    const props = createProps();

    const wrapper = shallow(<TimePicker {...props} />);

    const inputOnClick = wrapper.find('input').prop('onClick');
    inputOnClick();

    wrapper.update();

    expect(wrapper.state('isOpen')).toBe(true);
  });

  test('Can select hour', () => {
    const props = createProps();

    const wrapper = shallow(<TimePicker {...props} />);

    const inputOnClick = wrapper.find('input').prop('onClick');
    inputOnClick();

    wrapper.update();

    const date = new Date();
    date.setHours(0, date.getMinutes());

    const onClickFirstHour = wrapper
      .find('.popup')
      .find('.popupList')
      .first()
      .find('.popupItem')
      .first()
      .prop('onClick');

    onClickFirstHour();

    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toBeCalledWith(date);
  });

  test('Can select minute', () => {
    const props = createProps();

    const wrapper = shallow(<TimePicker {...props} />);

    const inputOnClick = wrapper.find('input').prop('onClick');
    inputOnClick();

    wrapper.update();

    const date = new Date();
    date.setHours(1, 0);

    const onClickFirstHour = wrapper
      .find('.popup')
      .find('.popupList')
      .at(1)
      .find('.popupItem')
      .first()
      .prop('onClick');

    onClickFirstHour();

    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toBeCalledWith(date);
  });

  test('Can select amPm', () => {
    const props = createProps();

    const wrapper = shallow(<TimePicker {...props} />);

    const inputOnClick = wrapper.find('input').prop('onClick');
    inputOnClick();

    wrapper.update();

    const date = new Date();
    date.setHours(13, date.getMinutes());

    const onClickFirstHour = wrapper
      .find('.popup')
      .find('.popupList')
      .at(2)
      .find('.popupItem')
      .last() // PM
      .prop('onClick');

    onClickFirstHour();

    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toBeCalledWith(date);
  });
});
