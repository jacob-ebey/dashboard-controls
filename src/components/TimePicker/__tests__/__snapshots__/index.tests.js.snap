// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TimePicker - index.js Raises onChange for valid input 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimePicker
    onChange={[Function]}
    value={1993-11-30T09:30:00.000Z}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="input  "
          onChange={[Function]}
          onClick={[Function]}
          value="1:45 PM"
        />,
        undefined,
        false,
      ],
      "className": "main",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "className": "input  ",
          "onChange": [Function],
          "onClick": [Function],
          "value": "1:45 PM",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      undefined,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="input  "
            onChange={[Function]}
            onClick={[Function]}
            value="1:45 PM"
          />,
          undefined,
          false,
        ],
        "className": "main",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "className": "input  ",
            "onChange": [Function],
            "onClick": [Function],
            "value": "1:45 PM",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        undefined,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`components/TimePicker - index.js Raises onChange if no value 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimePicker
    onChange={[Function]}
    value=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="input  "
          onChange={[Function]}
          onClick={[Function]}
          value=""
        />,
        undefined,
        false,
      ],
      "className": "main",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "className": "input  ",
          "onChange": [Function],
          "onClick": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      undefined,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="input  "
            onChange={[Function]}
            onClick={[Function]}
            value=""
          />,
          undefined,
          false,
        ],
        "className": "main",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "className": "input  ",
            "onChange": [Function],
            "onClick": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        undefined,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`components/TimePicker - index.js Renders without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimePicker
    onChange={[Function]}
    value={1993-11-30T09:30:00.000Z}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="input  "
          onChange={[Function]}
          onClick={[Function]}
          value="1:30 AM"
        />,
        undefined,
        false,
      ],
      "className": "main",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "className": "input  ",
          "onChange": [Function],
          "onClick": [Function],
          "value": "1:30 AM",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      undefined,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="input  "
            onChange={[Function]}
            onClick={[Function]}
            value="1:30 AM"
          />,
          undefined,
          false,
        ],
        "className": "main",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "className": "input  ",
            "onChange": [Function],
            "onClick": [Function],
            "value": "1:30 AM",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        undefined,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
